pandas/core/frame.py

```diff
@@ -2765,6 +2773,46 @@ def reindexer(value):
 
         return np.atleast_2d(np.asarray(value))
 
+    def _validate_axis_style_args(self, arg, arg_name, index, columns,
+                                  axis, method_name):
+        if axis is not None:
+            # Using "axis" style, along with a positional arg
```

Comment:
I suppose it would make it a lot harder if the default `axis=0` is used instead of None ?

Reply:
Right now I raise on 

```python
df.reindex(labels, axis=0)
df.reindex(index=labels, axis=0)
```

If I change the default axis to 0, then I can't detect those cases. I could go either way here.

Reply:
Why do you raise on `df.reindex(labels, axis=0)` ? That seems perfectly valid?

Reply:
It's valid, just the `axis=0` is redundant. Likewise with `columns=labels, axis=1` (I currently raise). Happy to adjust that though.

Reply:
that are two different things; because you can either specify mapper/axis or index/columns. 
`df.reindex(labels, axis=0)` is a case of mapper/axis, and is thus perfectly valid (and even more explicit than leaving out axis=0, although it is the default, so I don't think we should raise on this), while `df.reindex(columns=labels, axis=1)` is mixture of columns and axis, so OK to raise on that.

Reply:
`df.reindex(index=labels, axis=0)` is a bit the tricky case, as this one is indeed redundant and mixing the two. So in principle I would also raise here like for `df.reindex(columns=labels, axis=1)`. 
But if it is easier implementation-wise to allow that, I think that is OK (as although it is mixing both idioms, it is consistent in which axis compared to eg `df.reindex(index=labels, axis=1)`.

Reply:
Sorry, https://github.com/pandas-dev/pandas/pull/17800#discussion_r143211232 was incorrect. `.reindex(labels, axis=0)` should clearly work!

Reply:
```python
In [1]: import pandas as pd
d
In [2]: df = pd.DataFrame({"A": [1, 2, 3], "B": [4, 5, 6]})

In [3]: df.reindex([0, 1], axis=0)
Out[3]:
   A  B
0  1  4
1  2  5

In [4]: df.reindex(index=[0, 1], axis=0)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-d6f30fd70cc7> in <module>()
----> 1 df.reindex(index=[0, 1], axis=0)

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/frame.py in reindex(self, labels, index, columns, axis, **kwargs)
   2945         index, columns = self._validate_axis_style_args(labels, 'labels',
   2946                                                         index, columns,
-> 2947                                                         axis, 'reindex')
   2948         return super(DataFrame, self).reindex(index=index, columns=columns,
   2949                                               **kwargs)

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/frame.py in _validate_axis_style_args(self, arg, arg_name, index, columns, axis, method_name)
   2788                     "\t.{method_name}.rename(index=index, columns=columns)"
   2789                 ).format(arg_name=arg_name, method_name=method_name)
-> 2790                 raise TypeError(msg)
   2791             if axis == 'index':
   2792                 index = arg

TypeError: Can't specify both 'axis' and 'index' or 'columns'. Specify either
        .reindex.rename(labels, axis=axis), or
        .reindex.rename(index=index, columns=columns)

In [5]: df.reindex(columns=[0, 1], axis=1)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-02a868f92f79> in <module>()
----> 1 df.reindex(columns=[0, 1], axis=1)

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/frame.py in reindex(self, labels, index, columns, axis, **kwargs)
   2945         index, columns = self._validate_axis_style_args(labels, 'labels',
   2946                                                         index, columns,
-> 2947                                                         axis, 'reindex')
   2948         return super(DataFrame, self).reindex(index=index, columns=columns,
   2949                                               **kwargs)

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/frame.py in _validate_axis_style_args(self, arg, arg_name, index, columns, axis, method_name)
   2788                     "\t.{method_name}.rename(index=index, columns=columns)"
   2789                 ).format(arg_name=arg_name, method_name=method_name)
-> 2790                 raise TypeError(msg)
   2791             if axis == 'index':
   2792                 index = arg

TypeError: Can't specify both 'axis' and 'index' or 'columns'. Specify either
        .reindex.rename(labels, axis=axis), or
        .reindex.rename(index=index, columns=columns)

```